java/FileTest.java

```diff
@@ -0,0 +1,122 @@
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * Implements unit tests against test files (*_input.txt and corresponding *_output.txt)
+ * in the directory specified by the -Dtest.data variable.
+ * 
+ * @author Fabian Schliski
+ *
+ */
+@RunWith(Parameterized.class)
+public class FileTest {
+
+    String testName;
+    private String[] input;
+    private String expectedOutput;
+
+    public FileTest(String testName, String[] input, String expectedOutput) {
+        this.testName = testName;
+        this.input = input;
+        this.expectedOutput = expectedOutput;
+    }
+
+    @Parameters
+    public static Collection<Object[]> data() {
+        // Getting test.data property
+        final String testDataLocation = System.getProperty("test.data");
```

Comment:
What is `test.data`? The location where input and expected output files must be read? I am not sure that using system variables here is the right way to go. Maybe we can encode this value (like, ../test_data) or at least create an Assumption to validate the presence of this folder? I think that at least for the beginning, hardcoding the location of test data is a fair thing to do.

Reply:
Actually, since this seems to work (as you hardcoded the path in the make file) we can live with that.

Reply:
As I didn't want to hardcode it in the code I hardcoded it into the makefile (still better I guess?). And in the following lines I check whether the folder exists and is a folder, so it should be pretty safe.